src/WebJobs.Script.WebHost/Diagnostics/AzureMonitorDiagnosticLogger.cs

```diff
@@ -26,21 +26,24 @@ public class AzureMonitorDiagnosticLogger : ILogger
         private readonly string _hostInstanceId;
         private readonly string _roleInstance;
         private readonly HostNameProvider _hostNameProvider;
-        private readonly IOptionsMonitor<AppServiceOptions> _appServiceOptions;
         private readonly IEventGenerator _eventGenerator;
         private readonly IEnvironment _environment;
         private readonly IExternalScopeProvider _scopeProvider;
+        private AppServiceOptions _appServiceOptions;
 
         public AzureMonitorDiagnosticLogger(string category, string hostInstanceId, IEventGenerator eventGenerator, IEnvironment environment, IExternalScopeProvider scopeProvider,
-            HostNameProvider hostNameProvider, IOptionsMonitor<AppServiceOptions> appServiceOptions)
+            HostNameProvider hostNameProvider, IOptionsMonitor<AppServiceOptions> appServiceOptionsMonitor)
         {
             _category = category ?? throw new ArgumentNullException(nameof(category));
             _hostInstanceId = hostInstanceId ?? throw new ArgumentNullException(nameof(hostInstanceId));
             _eventGenerator = eventGenerator ?? throw new ArgumentNullException(nameof(eventGenerator));
             _environment = environment ?? throw new ArgumentNullException(nameof(environment));
             _scopeProvider = scopeProvider ?? throw new ArgumentNullException(nameof(scopeProvider));
             _hostNameProvider = hostNameProvider ?? throw new ArgumentNullException(nameof(hostNameProvider));
-            _appServiceOptions = appServiceOptions ?? throw new ArgumentNullException(nameof(appServiceOptions));
+            _ = appServiceOptionsMonitor ?? throw new ArgumentNullException(nameof(appServiceOptionsMonitor));
+
+            appServiceOptionsMonitor.OnChange(newOptions => _appServiceOptions = newOptions);
```

Comment:
How does the additional closure here compare to the allocations of the factory during the CurrentValue call?

Reply:
In this case we're talking about 1 closure allocation for the event -per logger_ (so, pretty much 1), whereas today we're talking about _per function execution_ which is an unbounded `n`. So indeed this isn't free, but it's way cheaper and only paid once :)